homeassistant/components/fan/__init__.py

```diff
@@ -105,19 +130,48 @@ async def async_unload_entry(hass, entry):
     return await hass.data[DOMAIN].async_unload_entry(entry)
 
 
+def _fan_native(method):
+    """Native fan method not overridden."""
+    setattr(method, _FAN_NATIVE, True)
+    return method
+
+
 class FanEntity(ToggleEntity):
     """Representation of a fan."""
 
+    @_fan_native
     def set_speed(self, speed: str) -> None:
         """Set the speed of the fan."""
         raise NotImplementedError()
 
+    @_fan_native
     async def async_set_speed(self, speed: str):
         """Set the speed of the fan."""
         if speed == SPEED_OFF:
             await self.async_turn_off()
+        elif not hasattr(self.async_set_percentage, _FAN_NATIVE):
+            await self.async_set_percentage(self.speed_to_percentage(speed))
```

Comment:
We may need a check here since `speed_to_percentage` may return `None`.

Reply:
I'm not sure if its better to raise a named exception or let it blow up since I don't think there is any way to know what the user wants since *I think* the only way this happens is if the fan has no valid speeds.

Reply:
I think we should raise already in `speed_to_percentage` and consolidate the return type.